app/controllers/dances_controller.rb

```diff
@@ -85,8 +85,14 @@ def authenticate_dance_writable!
 
     def authenticate_dance_readable!
       unless @dance.readable?(current_user)
-        flash[:notice] = "this dance has not been published"
-        redirect_back(fallback_location: '/dances')
+        if current_user
+          flash[:notice] = "that dance has not been published so you can't see it"
+          redirect_back(fallback_location: root_path)
+        else
+          flash[:notice] = "that dance has not been published - maybe it's yours and you could see it if you logged in?"
+          session[:after_login] = request.env['PATH_INFO']
+          redirect_to(new_user_session_path)
```

Comment:
Could these two lines be profitably split into a helper function? If so, could we use something similar on the New Dance button?